addons/udes_security/controllers/redirect.py

```diff
@@ -1,11 +1,72 @@
 import logging
-from odoo import http
+from odoo import http, _
 from odoo.addons.web.controllers import main
 from odoo.http import request
+from werkzeug import urls, url_unquote
+from werkzeug.exceptions import BadRequest
+from fnmatch import fnmatch
 
 _logger = logging.getLogger(__name__)
 
 
+class no_absolute_redirect(object):
+    """Decorator to mark a route as forbidding absolute redirects
+
+    If the decorated route is called with a "redirect" query parameter
+    containing an absolute URL, the request will be terminated with an
+    exception.
+    """
+
+    def __init__(self, func):
+        self.func = func
+
+    @property
+    def __name__(self):
+        """Copy original method name
+
+        This is required since the method name will be introspected as
+        part of calculating the controller inheritance structure.
+        """
+        return self.func.__name__
+
+    def __call__(self, *args, **kwargs):
+        """Call method"""
+        redirect = kwargs.get("redirect")
+
+        if redirect:
+            redirect = url_unquote(redirect)
+            if redirect.startswith("//"):
+                # remove leading /'s which can trick the url parser but is handled by browsers
+                kwargs["redirect"] = "/" + redirect.lstrip("/")
+            else:
+                parsed_url = urls.url_parse(redirect)
+
+                if self.is_absolute_url(parsed_url) and not self.domain_is_allowlisted(parsed_url):
+                    _logger.error("Invalid redirection attempted: %s", redirect)
+                    raise BadRequest(_("Invalid redirection attempted"))
+
+        return self.func(*args, **kwargs)
+
+    @staticmethod
+    def is_absolute_url(parsed_url):
+        """Test if a URL is an absolute URL"""
```

Comment:
Super nit: Remove the term `test` from the doc-string here and for `domain_is_allowlisted`.

E.g. `Returns True if URL is an absolute URL, else False` or similar.